dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ProcessingManager.java

```diff
@@ -54,6 +54,7 @@ private ProcessingManager(Configuration config) throws InvalidConfigurationExcep
                 constantPropagation ? ConstantPropagation.fromConfig(config) : null,
                 DeadAssignmentElimination.fromConfig(config),
                 RemoveDeadCondJumps.fromConfig(config),
+                DeadCodeElimination.fromConfig(config),
```

Comment:
A dead register reader could have an empty list of writers, violating some assertions in `Dependency`.  A dead register writer could leave its reference behind, which would require more cleanup.

Reply:
If possible, I would keep the preconditions as few as possible. If there is a way to treat dead reads/writes without much issues, I think that is a better solution. If it is hard to do and the DCE is really needed, we should definitely put some comment on the class that lists this requirement (maybe you already did, I didn't check).

Reply:
This line was a temporary fix for RemoveDeadCondJumps introducing new dead code.  As mentioned in #219, DCE is meant to be run before unrolling.  Therefore, i'll add a small DCE to `RemoveDeadCondJumps`, such that it cleans up after itself.

Reply:
There is something that it is not clear to me yet (the first comment by @xeren was supposed to answer this, but I did not really get it) ... do we really need the first DCE pass? Can we just keep the second DCE pass (the one before unrolling) so it takes care of cleaning any dead code another pass may introduce?

Reply:
If the DCE before `BranchReordering` is removed, branch reordering will order unreachable code to before `Thread.getEntry` resulting in pseudo-deleted code. It is deleted in the sense that `Thread.getEvents` won't return that code anymore because it is unreachable starting from `Thread.getEntry`. However, it does not get garbage collected because we keep (unintended) references to the dead code via e.g. `Label.getListeners`. 
Overall, this can cause memory leaks. A subsequent run of `DCE` cannot delete that code because it is unreachable via `Event.getSuccessor`.
In short: the first `DCE` is crucial as long as we use `BranchReordering`

Reply:
This change should have been undone.  Instead, `RemoveDeadCondJumps` now performs its own DCE.

Reply:
You mean, `Dependency` should not assert that certain processes like DCE were performed beforehand, such that they still can be opted out when desired.  There is still one assertion that has to stay for now: That the thread that calls `pthread_create` appears before the created thread in the list of threads.

Reply:
Yes, I think it is best if the preconditions under which an analysis or a program transformation runs should be as small as possible. If handling dead code requires a couple of lines of extra code but removes the precondition of having all dead code eliminated, I would rather add the few extra lines of code to buy flexibility. And if the analysis just becomes inaccurate without proper preprocessing then so be it. In that case it is our fault for not doing the preprocessing beforehand.
If the analysis is simply not possible without specific preprocessing steps, then that is fine and we can require those steps beforehand.
I believe some of the existing analysis/preprocessors have unnecessary requirements (like DCE being done before unrolling) and we should eventually remove those as well.

Reply:
Good point.  After I have added some logging, I think this PR will be ready to be merged.